lib/vfscore: Fix write const path in automount

andreittr opened the pull request:
### Description of changes

Previously vfscore_mount_volume would pass a const path to vfscore_ukopt_mkmp, which expects a mutable path, and indeed does modify it in-place during execution. This is wrong and rightfully triggers a compiler warning.
This change fixes the issue, by passing a temporary copy to mkmp.

### Prerequisite checklist

 - [x] Read the [contribution guidelines](https://unikraft.org/docs/contributing/) regarding submitting new changes to the project;
 - [x] Tested your changes against relevant architectures and platforms;
 - [x] Ran the [`checkpatch.uk`](https://github.com/unikraft/unikraft/blob/staging/support/scripts/checkpatch.uk) on your commit series before opening this PR;
 - [ ] Updated relevant documentation.


### Base target

 - Architecture(s): [e.g. `x86_64` or N/A]
 - Platform(s): [e.g. `kvm`, `xen` or N/A]
 - Application(s): [e.g. `app-python3` or N/A]


### Additional configuration

N/A

michpappas commented on lib/vfscore/automount.c:
```diff
@@ -687,7 +687,10 @@ static inline int vfscore_mount_volume(const struct vfscore_volume *vv)
 
 	if (vv->ukopts & LIBVFSCORE_UKOPT_MKMP) {
-		rc = vfscore_ukopt_mkmp(path);
+		char tmp[strlen(path) + 1];
```
Not a blocker, but FYI: Similarly to Linux, our convention is not to use scope declarations, especially given that we don't yet have consensus on the standard the code should comply to.

andreittr replied:
> our convention is not to use scope declarations

That is news to me, as it's not mentioned in the [coding conventions](https://unikraft.org/docs/contributing/coding-conventions#variable-definitions). `tmp`  is indeed declared at the start of a block and separated by whitespace from code. But fine, I can move `tmp` and the `strcpy` to `mkmp`, doesn't make much of a difference for this case. For the record, I _absolutely_ disagree that "scope declarations are bad and we should avoid them", but that debate is out of scope.

> we don't yet have consensus on the standard the code should comply to

The core codebase uses C99 features extensively (whether by design or not, this is where we stand atm); Unikraft supporting C89/90 (which cannot handle intermingled declarations & code, and a major driver of most of Linux's very outdated codestyle) is IMHO impossible. And even C89 supports declarations at the beginning of any block, not just function scope. Thus I believe any concern about functionality/correctness is unfounded; this is only a question of style.

michpappas replied:
> That is news to me, as it's not mentioned in the [coding conventions](https://unikraft.org/docs/contributing/coding-conventions#variable-definitions). tmp is indeed declared at the start of a block and separated by whitespace from code. But fine, I can move tmp and the strcpy to mkmp, doesn't make much of a difference for this case. For the record, I absolutely disagree that "scope declarations are bad and we should avoid them", but that debate is out of scope.

Indeed neither in our or in the linux's conventions it's not explicit. And for the record in principle I agree with you as I would also think they can be useful at least in some cases, but we cannot apply the conventions selectively depending to our preference. So as you suggest, let's have that discussion, but not in the scope of this PR 👍🏼 

Similarly for the topic of C99 (agreed, let's do it, not in this scope).

michpappas commented on lib/vfscore/automount.c:
```diff
@@ -687,7 +687,10 @@ static inline int vfscore_mount_volume(const struct vfscore_volume *vv)
+		char tmp[strlen(path) + 1];
+
+		strcpy(tmp, path);
+		rc = vfscore_ukopt_mkmp(tmp);
```
Looks like `vfscore_ukopt_mkmp()` modifies the temp path, while later we pass `path` to `mount()`. Is it safe to do so?

andreittr replied:
Yes; the modifications mkmp does to the path received as arg are undone by the time it exits. But it's still _writing_ to that memory, so a `const char *` is inappropriate.

michpappas replied:
Thanks for confirming

michpappas commented:
Also, as checkpatch suggests, the commit title should have a single scope component, i.e. `lib/vfscore: Fix write const path in automount`.

andreittr commented:
> Also, as checkpatch suggests, the commit title should have a single scope component, i.e. lib/vfscore: Fix write const path in automount.

That had been my initial title (w/ "write _of_"), but it overflows git's allowed/recommented title limit. Will fix to remove "of" & it fits.

andreittr committed:
lib/vfscore: Fix write const path in automount

Previously vfscore_mount_volume would pass a const path to
vfscore_ukopt_mkmp, which expects a mutable path, and indeed does modify
it in-place during execution. This is wrong and rightfully triggers a
compiler warning.
This change fixes this by having mkmp allocate a temporary writable path.

Checkpatch-Ignore: STRCPY
Signed-off-by: Andrei Tatar <andrei@unikraft.io>

michpappas commented:
Reviewed-by: Michalis Pappas <michalis@unikraft.io>

michpappas commented:
Approved-by: Michalis Pappas <michalis@unikraft.io>
